File: packages/SystemUI/src/com/android/systemui/biometrics/AuthBiometricFaceIconController.java
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.systemui.biometrics;

import android.content.Context;
import android.graphics.drawable.Animatable2;
import android.graphics.drawable.AnimatedVectorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.ImageView;
import android.widget.TextView;

import com.android.systemui.R;

class AuthBiometricFaceIconController extends Animatable2.AnimationCallback {

    private static final String TAG = "AuthBiometricFaceIconController";

    protected Context mContext;
    protected ImageView mIconView;
    protected TextView mTextView;
    protected Handler mHandler;
    protected boolean mLastPulseLightToDark; // false = dark to light, true = light to dark
    @AuthBiometricView.BiometricState protected int mState;
    protected boolean mDeactivated;

    protected AuthBiometricFaceIconController(Context context, ImageView iconView,
            TextView textView) {
        mContext = context;
        mIconView = iconView;
        mTextView = textView;
        mHandler = new Handler(Looper.getMainLooper());
        showStaticDrawable(R.drawable.face_dialog_pulse_dark_to_light);
    }

    protected void animateOnce(int iconRes) {
        animateIcon(iconRes, false);
    }

    protected void showStaticDrawable(int iconRes) {
        mIconView.setImageDrawable(mContext.getDrawable(iconRes));
    }

    protected void animateIcon(int iconRes, boolean repeat) {
        Log.d(TAG, "animateIcon, state: " + mState + ", deactivated: " + mDeactivated);
        if (mDeactivated) {
            return;
        }

        final AnimatedVectorDrawable icon =
                (AnimatedVectorDrawable) mContext.getDrawable(iconRes);
        mIconView.setImageDrawable(icon);
        icon.forceAnimationOnUI();
        if (repeat) {
            icon.registerAnimationCallback(this);
        }
        icon.start();
    }

    protected void startPulsing() {
        mLastPulseLightToDark = false;
        animateIcon(R.drawable.face_dialog_pulse_dark_to_light, true);
    }

    protected void pulseInNextDirection() {
        int iconRes = mLastPulseLightToDark ? R.drawable.face_dialog_pulse_dark_to_light
                : R.drawable.face_dialog_pulse_light_to_dark;
        animateIcon(iconRes, true /* repeat */);
        mLastPulseLightToDark = !mLastPulseLightToDark;
    }

    @Override
    public void onAnimationEnd(Drawable drawable) {
        super.onAnimationEnd(drawable);
        Log.d(TAG, "onAnimationEnd, mState: " + mState + ", deactivated: " + mDeactivated);
        if (mDeactivated) {
            return;
        }

        if (mState == AuthBiometricView.STATE_AUTHENTICATING
                || mState == AuthBiometricView.STATE_HELP) {
            pulseInNextDirection();
        }
    }

    protected void deactivate() {
        mDeactivated = true;
    }

    protected void updateState(int lastState, int newState) {
        if (mDeactivated) {
            Log.w(TAG, "Ignoring updateState when deactivated: " + newState);
            return;
        }

        final boolean lastStateIsErrorIcon =
                lastState == AuthBiometricView.STATE_ERROR
                        || lastState == AuthBiometricView.STATE_HELP;

        if (newState == AuthBiometricView.STATE_AUTHENTICATING_ANIMATING_IN) {
            showStaticDrawable(R.drawable.face_dialog_pulse_dark_to_light);
            mIconView.setContentDescription(mContext.getString(
                    R.string.biometric_dialog_face_icon_description_authenticating));
        } else if (newState == AuthBiometricView.STATE_AUTHENTICATING) {
            startPulsing();
            mIconView.setContentDescription(mContext.getString(
                    R.string.biometric_dialog_face_icon_description_authenticating));
        } else if (lastState == AuthBiometricView.STATE_PENDING_CONFIRMATION
                && newState == AuthBiometricView.STATE_AUTHENTICATED) {
            animateOnce(R.drawable.face_dialog_dark_to_checkmark);
            mIconView.setContentDescription(mContext.getString(
                    R.string.biometric_dialog_face_icon_description_confirmed));
        } else if (lastStateIsErrorIcon && newState == AuthBiometricView.STATE_IDLE) {
            animateOnce(R.drawable.face_dialog_error_to_idle);
            mIconView.setContentDescription(mContext.getString(
                    R.string.biometric_dialog_face_icon_description_idle));
        } else if (lastStateIsErrorIcon && newState == AuthBiometricView.STATE_AUTHENTICATED) {
            animateOnce(R.drawable.face_dialog_dark_to_checkmark);
            mIconView.setContentDescription(mContext.getString(
                    R.string.biometric_dialog_face_icon_description_authenticated));
        } else if (newState == AuthBiometricView.STATE_ERROR
                && lastState != AuthBiometricView.STATE_ERROR) {
            animateOnce(R.drawable.face_dialog_dark_to_error);
        } else if (lastState == AuthBiometricView.STATE_AUTHENTICATING
                && newState == AuthBiometricView.STATE_AUTHENTICATED) {
            animateOnce(R.drawable.face_dialog_dark_to_checkmark);
            mIconView.setContentDescription(mContext.getString(
                    R.string.biometric_dialog_face_icon_description_authenticated));
        } else if (newState == AuthBiometricView.STATE_PENDING_CONFIRMATION) {
            animateOnce(R.drawable.face_dialog_wink_from_dark);
            mIconView.setContentDescription(mContext.getString(
                    R.string.biometric_dialog_face_icon_description_authenticated));
        } else if (newState == AuthBiometricView.STATE_IDLE) {
            showStaticDrawable(R.drawable.face_dialog_idle_static);
            mIconView.setContentDescription(mContext.getString(
                    R.string.biometric_dialog_face_icon_description_idle));
        } else {
            Log.w(TAG, "Unhandled state: " + newState);
        }
        mState = newState;
    }
}

File: packages/SystemUI/src/com/android/systemui/biometrics/AuthBiometricFaceIconController.kt
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.systemui.biometrics

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.Log
import com.airbnb.lottie.LottieAnimationView
import com.android.systemui.res.R
import com.android.systemui.biometrics.ui.binder.Spaghetti.BiometricState

private const val TAG = "AuthBiometricFaceIconController"

/** Face only icon animator for BiometricPrompt. */
class AuthBiometricFaceIconController(
        context: Context,
        iconView: LottieAnimationView
) : AuthIconController(context, iconView) {

    // false = dark to light, true = light to dark
    private var lastPulseLightToDark = false

    private var state: BiometricState = BiometricState.STATE_IDLE

    init {
        val size = context.resources.getDimensionPixelSize(R.dimen.biometric_dialog_face_icon_size)
        iconView.layoutParams.width = size
        iconView.layoutParams.height = size
        showStaticDrawable(R.drawable.face_dialog_pulse_dark_to_light)
    }

    private fun startPulsing() {
        lastPulseLightToDark = false
        animateIcon(R.drawable.face_dialog_pulse_dark_to_light, true)
    }

    private fun pulseInNextDirection() {
        val iconRes = if (lastPulseLightToDark) {
            R.drawable.face_dialog_pulse_dark_to_light
        } else {
            R.drawable.face_dialog_pulse_light_to_dark
        }
        animateIcon(iconRes, true /* repeat */)
        lastPulseLightToDark = !lastPulseLightToDark
    }

    override fun handleAnimationEnd(drawable: Drawable) {
        if (state == BiometricState.STATE_AUTHENTICATING || state == BiometricState.STATE_HELP) {
            pulseInNextDirection()
        }
    }

    override fun updateIcon(oldState: BiometricState, newState: BiometricState) {
        val lastStateIsErrorIcon = (oldState == BiometricState.STATE_ERROR || oldState == BiometricState.STATE_HELP)
        if (newState == BiometricState.STATE_AUTHENTICATING_ANIMATING_IN) {
            showStaticDrawable(R.drawable.face_dialog_pulse_dark_to_light)
            iconView.contentDescription = context.getString(
                    R.string.biometric_dialog_face_icon_description_authenticating
            )
        } else if (newState == BiometricState.STATE_AUTHENTICATING) {
            startPulsing()
            iconView.contentDescription = context.getString(
                    R.string.biometric_dialog_face_icon_description_authenticating
            )
        } else if (oldState == BiometricState.STATE_PENDING_CONFIRMATION && newState == BiometricState.STATE_AUTHENTICATED) {
            animateIconOnce(R.drawable.face_dialog_dark_to_checkmark)
            iconView.contentDescription = context.getString(
                    R.string.biometric_dialog_face_icon_description_confirmed
            )
        } else if (lastStateIsErrorIcon && newState == BiometricState.STATE_IDLE) {
            animateIconOnce(R.drawable.face_dialog_error_to_idle)
            iconView.contentDescription = context.getString(
                    R.string.biometric_dialog_face_icon_description_idle
            )
        } else if (lastStateIsErrorIcon && newState == BiometricState.STATE_AUTHENTICATED) {
            animateIconOnce(R.drawable.face_dialog_dark_to_checkmark)
            iconView.contentDescription = context.getString(
                    R.string.biometric_dialog_face_icon_description_authenticated
            )
        } else if (newState == BiometricState.STATE_ERROR && oldState != BiometricState.STATE_ERROR) {
            animateIconOnce(R.drawable.face_dialog_dark_to_error)
            iconView.contentDescription = context.getString(
                    R.string.keyguard_face_failed
            )
        } else if (oldState == BiometricState.STATE_AUTHENTICATING && newState == BiometricState.STATE_AUTHENTICATED) {
            animateIconOnce(R.drawable.face_dialog_dark_to_checkmark)
            iconView.contentDescription = context.getString(
                    R.string.biometric_dialog_face_icon_description_authenticated
            )
        } else if (newState == BiometricState.STATE_PENDING_CONFIRMATION) {
            animateIconOnce(R.drawable.face_dialog_wink_from_dark)
            iconView.contentDescription = context.getString(
                    R.string.biometric_dialog_face_icon_description_authenticated
            )
        } else if (newState == BiometricState.STATE_IDLE) {
            showStaticDrawable(R.drawable.face_dialog_idle_static)
            iconView.contentDescription = context.getString(
                    R.string.biometric_dialog_face_icon_description_idle
            )
        } else {
            Log.w(TAG, "Unhandled state: $newState")
        }
        state = newState
    }
}
